src/headers/tomcrypt_misc.h

```diff
@@ -154,6 +154,25 @@ int padding_pad(unsigned char *data, unsigned long length, unsigned long* padded
 int padding_depad(const unsigned char *data, unsigned long *length, unsigned long mode);
 #endif  /* LTC_PADDING */
 
+#ifdef LTC_SSH
+typedef enum ssh_data_type_ {
+   LTC_SSHDATA_BYTE,
+   LTC_SSHDATA_BOOLEAN,
+   LTC_SSHDATA_UINT32,
+   LTC_SSHDATA_UINT64,
+   LTC_SSHDATA_STRING,
+   LTC_SSHDATA_MPINT,
+   LTC_SSHDATA_NAMELIST,
+   LTC_SSHDATA_EOL
+} ssh_data_type;
+
+int ssh_test(void);
+
+/* VA list handy helpers with tuples of <type, data> */
+int ssh_encode_sequence_multi(unsigned char *out, unsigned long *outlen, ...);
```

Comment:
Not sure how well va_args lists handle restarting in the middle of the fixed parameters?

I'm going to alter the implementation so we don't have `void *data` being cast all over the place but explicitly collect the correct type from the stack.

Reply:
how about something like:

```
diff --git a/src/misc/ssh/ssh_decode_sequence_multi.c b/src/misc/ssh/ssh_decode_sequence_multi.c
index 2507609..3ace026 100644
--- a/src/misc/ssh/ssh_decode_sequence_multi.c
+++ b/src/misc/ssh/ssh_decode_sequence_multi.c
@@ -25,7 +25,7 @@
 */
-int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
+int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, int type, void *data, ...)
 {
    int           err;
-   ssh_data_type type;
-   void          *data;
+   ssh_data_type t;
+   void          *d;
    va_list       args;
@@ -36,7 +36,6 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
    /* Decode values from buffer */
-   va_start(args, inlen);
+   va_start(args, data);
+   t = type;
+   d = data;
    for (;;) {
-      type = (ssh_data_type)va_arg(args, int);
-      data = va_arg(args, void*);
-      LTC_UNUSED_PARAM(data);
 
@@ -46,3 +45,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
 
-      switch (type) {
+      switch (t) {
          case LTC_SSHDATA_BYTE:
@@ -122,6 +121,9 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
          case LTC_SSHDATA_EOL: /* Should never get here */
-            va_end(args);
-            return CRYPT_INVALID_ARG;
+            err = CRYPT_INVALID_ARG;
+            goto out;
        }
+      t = (ssh_data_type)va_arg(args, int);
+      d = va_arg(args, void*);
    }
+out:
    va_end(args);
```

Reply:
I've gone for something a bit more radical... just updating the tests to cover byte/boolean/uint64 now. Let me know what you think, I can always change it if you're not happy.

Reply:
Changed to a single return path and re-pushed.

Reply:
Ah... `warning: second argument to 'va_start' is not the last named parameter [-Wvarargs]`. Since it has to go through the loop twice, `type` can't be in the formal parameters for `ssh_encode_sequence_multi`...

Reply:
I guess you forgot to update the `va_start()` argument...

```
diff --git a/src/misc/ssh/ssh_decode_sequence_multi.c b/src/misc/ssh/ssh_decode_sequence_multi.c
index ca088f9..e44162a 100644
--- a/src/misc/ssh/ssh_decode_sequence_multi.c
+++ b/src/misc/ssh/ssh_decode_sequence_multi.c
@@ -25,6 +25,6 @@
 */
-int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
+int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, int type_, ...)
 {
    int           err;
-   ssh_data_type type;
+   ssh_data_type type = type_;
    va_list       args;
@@ -34,5 +34,4 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
    /* Decode values from buffer */
-   va_start(args, inlen);
+   va_start(args, type_);
    for (;;) {
-      type = (ssh_data_type)va_arg(args, int);
 
@@ -103,2 +102,3 @@ int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
       }
+      type = (ssh_data_type)va_arg(args, int);
    }
```
this compiles w/o warning and tests are OK
(please don't call it `type_` ;) )

Reply:
That's different to how `der_encode_sequence_multi` etc handle varargs, but I can do it that way if you want?

Reply:
OK, leave the signatures as they are, sorry for the noise :)

Reply:
No worries... it's annoying that there's no elegant solution, it's like an itch in the brain! :)